apps/bc/contracts/BC.sol

```diff
@@ -0,0 +1,353 @@
+    function getPricePPM(address collateralToken, uint256 _totalSupply, uint256 _poolBalance) public view returns (uint256) {
+        return uint256(ppm).mul(_poolBalance) / _totalSupply.mul(reserveRatios[collateralToken]);
+        // return getSlope(_totalSupply, _poolBalance).mul(_totalSupply ** getPolynomial()) / ppm;
+    }
+    function getBuy(address collateralToken, uint256 _totalSupply, uint256 _poolBalance, uint256 buyValue) public view returns (uint256) {
+        return calculatePurchaseReturn(
+            safeAdd(_totalSupply, virtualSupplies[collateralToken]), 
+            safeAdd(_poolBalance, virtualBalances[collateralToken]), 
+            reserveRatios[collateralToken], 
+            buyValue);
+    }
+    function getSell(address collateralToken, uint256 _totalSupply, uint256 _poolBalance, uint256 sellAmount) public view returns (uint256) {
+        return calculateSaleReturn(safeAdd(_totalSupply, virtualSupplies[collateralToken]), safeAdd(_poolBalance, virtualBalances[collateralToken]), reserveRatios[collateralToken], sellAmount);
+    }
+
+    // totalSupply remains the same
+    // balance remains the same (although collateral has been collected)
+    function addBuy(address collateralToken, uint256 value, address sender) public returns (bool) {
+        uint256 batch = currentBatch();
+        Batch storage cb = batchesByCollateralToken[collateralToken][batch]; // currentBatch
+        if (!cb.init) {
+            _initBatch(batch);
+        }
+        require(ERC20(collateralToken).transferFrom(msg.sender, address(this), value), ERROR_TRANSFERFROM_FAILED);
+        cb.totalBuySpend = cb.totalBuySpend.add(value);
+        if (cb.buyers[sender] == 0) {
+            addressToBlocksByCollateralToken[collateralToken][sender].push(batch);
+        }
+        cb.buyers[sender] = cb.buyers[sender].add(value);
+        return true;
+    }
+    // totalSupply is decremented
+    // balance remains the same
+    function addSell(address collateralToken, uint256 amount) public returns (bool) {
+        require(ERC20(bondingCurveToken).balanceOf(msg.sender) >= amount, "insufficient funds to do that");
+        uint256 batch = currentBatch();
+        Batch storage cb = batchesByCollateralToken[collateralToken][batch]; // currentBatch
+        if (!cb.init) {
+            _initBatch(batch);
+        }
+        cb.totalSellSpend = cb.totalSellSpend.add(amount);
+        if (cb.sellers[msg.sender] == 0) {
+            addressToBlocksByCollateralToken[collateralToken][msg.sender].push(batch);
+        }
+        cb.sellers[msg.sender] = cb.sellers[msg.sender].add(amount);
+        tokenManager.burn(msg.sender, amount);
+        return true;
+    }
+
+    function clearBatches() public {
+        for (uint256 i = 0; i < collateralTokens.length; i++) {
+            address collateralToken = collateralTokens[i];
+            _clearBatch(collateralToken);
+        }
+    }
+    /***** internal functions *****/
```

Comment:
For all these internal functions - which actually implement the contract logic - we should add event emitting. Events reducing is how Aragon apps frontend keep their states up-to-date [à la drizzle]. So If we emit no events, we're fucked :)